runtime/jcl/common/stdinit.c

```diff
@@ -453,14 +453,24 @@ internalInitializeJavaLangClassLoader(JNIEnv * env)
 
 	vmFuncs->internalEnterVMFromJNI(vmThread);
 
-	vm->applicationClassLoader = J9VMJAVALANGCLASSLOADER_VMREF(vmThread, J9_JNI_UNWRAP_REFERENCE(appClassLoader));
+	/* Always use the persisted applicationClassLoader in restore runs. */
+	if (!IS_RESTORE_RUN(vm)) {
+		vm->applicationClassLoader = J9VMJAVALANGCLASSLOADER_VMREF(vmThread, J9_JNI_UNWRAP_REFERENCE(appClassLoader));
+	}
 
 	if (NULL == vm->applicationClassLoader) {
 		/* CMVC 201518
 		 * applicationClassLoader may be null due to lazy classloader initialization. Initialize
 		 * the applicationClassLoader now or vm will start throwing NoClassDefFoundException.
 		 */
-		vm->applicationClassLoader = (void*) (UDATA)(vmFuncs->internalAllocateClassLoader(vm, J9_JNI_UNWRAP_REFERENCE(appClassLoader)));
+		if (!IS_RESTORE_RUN(vm)) {
+			vm->applicationClassLoader = (void *)(UDATA)(vmFuncs->internalAllocateClassLoader(vm, J9_JNI_UNWRAP_REFERENCE(appClassLoader)));
+#if defined(J9VM_OPT_SNAPSHOTS)
+		} else {
+			vmFuncs->initializeSnapshotClassLoaderObject(vm, vm->applicationClassLoader, J9_JNI_UNWRAP_REFERENCE(appClassLoader));
+#endif /* defined(J9VM_OPT_SNAPSHOTS) */
```

Comment:
Again, I prefer the earlier version of this.

Reply:
In this file, there are a few other unguarded `if (!IS_RESTORE_RUN) {` based on your previous comments. I'd flipped this `if`-`else` to keep the `!IS_RESTORE_RUN` case in the `if` block. But I agree that in the cases where we have an `else`, it makes the most sense to keep the guards around the `IS_X_RUN` code.

Reply:
My concern is the possible misinterpretation by readers of that this code might do nothing if `IS_RESTORE_RUN(vm)`; a mistake that could not be made with an earlier version:
```c
#if defined(J9VM_OPT_SNAPSHOTS)
		if (IS_RESTORE_RUN(vm)) {
			vmFuncs->initializeSnapshotClassLoaderObject(vm, vm->applicationClassLoader, J9_JNI_UNWRAP_REFERENCE(appClassLoader));
		} else
#endif /* defined(J9VM_OPT_SNAPSHOTS) */
		{
			vm->applicationClassLoader = (void *)(UDATA)(vmFuncs->internalAllocateClassLoader(vm, J9_JNI_UNWRAP_REFERENCE(appClassLoader)));
		}
```